### PR TITLE
feat: add transformers to flatten single field types automatically

### LINKED ISSUE #2343: transformer to flatten out types with single field
Write a `config/transformer` that unwraps types with one field.

**FROM**
```graphql
type Query {
  foo: Foo
}

# Type with only one field
type Foo { 
  bar: Bar
}

# Type with only one field
type Bar {
  a: Int
}
```

**TO**

```graphql
type Query {
  foo: Int
}
```

This can be achieved by using the `addField` and `omit` operators. 

**Technical Requirements**
- Should be configurable via the `Preset` settings.
- Should be enabled by default in config generation.
- Should be added before BetterTypeName but after TreeShaking transformers.
- Add integration tests in `tests/execution`

**IMPORTANT**
- The "Before" is the server schema and the "After" is the expected client schema.
- Link:
	- https://tailcall.run/docs/tailcall-dsl-graphql-custom-directives/#addfield-directive 
	- https://tailcall.run/docs/tailcall-dsl-graphql-custom-directives/#modify-directive
	- https://tailcall.run/docs/contributors/integration-testing/

### PR DESCRIPTION
**Pull request by Shylock-Hg**
**Summary:**  
Flatten single field by addField
**Issue Reference(s):**  
Close #2343 
/claim #2343 

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

File: src/core/config/transformer/snapshots/tailcall__core__config__transformer__flatten_single_field__test__type_name_generator_transform.snap [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+type Foo {
+  bar: Bar
+}
+
+type NotSingle {
+  f1: Int
+  f2: Int
+}
+
+type Query @addField(name: "foo", path: ["foo", "bar", "a"]) {
```
Comment (tusharmath) [inline src/core/config/transformer/snapshots/tailcall__core__config__transformer__flatten_single_field__test__type_name_generator_transform.snap:23]: Can you add some integration tests also.

Comment (tusharmath) [reply src/core/config/transformer/snapshots/tailcall__core__config__transformer__flatten_single_field__test__type_name_generator_transform.snap:23]: https://tailcall.run/docs/contributors/integration-testing/

**Comment by tusharmath** [comment]
Moving to draft to reduce noise and improve CI efficiency. Once you are ready just mark it as "ready to review". Feel free to give a shoutout on the #contributors channel on Discord if you want immediate attention.

**Review by BenNeighbour — APPROVED** [review]

**Comment by Shylock-Hg** [comment]
@tusharmath Please take a look.

File: src/core/config/transformer/flatten_single_field.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+/// Flat single field type and inline to Query directly by addField
+#[derive(Default)]
+pub struct FlattenSingleField;
+
+fn get_single_field_path(
+    config: &Config,
+    field_name: &str,
+    type_name: &str,
+) -> Option<Vec<String>> {
```
Comment (tusharmath) [inline src/core/config/transformer/flatten_single_field.rs:13]: Could lead to infinite loops.

File: tailcall-fixtures/fixtures/configs/flatten_single_field.graphql [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+# Type with only one field
+type Bar {
+  a: Int
+}
+
+type NotSingle {
+  f1: Int
+  f2: Int
+}
```
Comment (tusharmath) [inline tailcall-fixtures/fixtures/configs/flatten_single_field.graphql:23]: Add more tests With some recursive types, and more complex nesting.

NOTE: Ensure the test case remains easy to understand no matter how complex the nesting is.

**Comment by tusharmath** [comment]
Moving to draft to reduce noise and improve CI efficiency. Once you are ready just mark it as "ready to review". Feel free to give a shoutout on the #contributors channel on Discord if you want immediate attention.

File: src/core/config/transformer/snapshots/tailcall__core__config__transformer__flatten_single_field__test__type_name_generator_transform.snap (hunk 1 above)
Comment (tusharmath) [reply src/core/config/transformer/snapshots/tailcall__core__config__transformer__flatten_single_field__test__type_name_generator_transform.snap:23]: You are right ... I don't think we need those.

File: tailcall-fixtures/fixtures/configs/flatten_single_field.graphql (hunk 3 above)
Comment (Shylock-Hg) [reply tailcall-fixtures/fixtures/configs/flatten_single_field.graphql:25]: Ok.

File: src/core/config/transformer/flatten_single_field.rs (hunk 2 above)
Comment (Shylock-Hg) [reply src/core/config/transformer/flatten_single_field.rs:16]: Good catch. It's done.

**Review by tusharmath — APPROVED** [review]

**Pull request merged**